pkg/chunk/repair_test.go

```diff
@@ -0,0 +1,296 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package chunk_test
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	accountingmock "github.com/ethersphere/bee/pkg/accounting/mock"
+	"github.com/ethersphere/bee/pkg/api"
+	"github.com/ethersphere/bee/pkg/chunk"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/netstore"
+	"github.com/ethersphere/bee/pkg/p2p/streamtest"
+	"github.com/ethersphere/bee/pkg/pss"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	pushsyncmock "github.com/ethersphere/bee/pkg/pushsync/mock"
+	"github.com/ethersphere/bee/pkg/retrieval"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	storemock "github.com/ethersphere/bee/pkg/storage/mock"
+	chunktesting "github.com/ethersphere/bee/pkg/storage/testing"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+// TestRecoveryHook tests that a recovery hook can be created and called
+func TestRecoveryHook(t *testing.T) {
+	// test variables needed to be correctly set for any recovery hook to reach the sender func
+	chunkAddr := chunktesting.GenerateTestRandomChunk().Address()
+	target := "0xedC69a0F0E81394bb08F90F89e35F93287E99dc1"
+	ctx := context.WithValue(context.Background(), api.TargetsContextKey{}, target)
+
+	// setup the sender
+	hookWasCalled := false // test variable to check if hook is called
+	testSender := func(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+		hookWasCalled = true
+		return nil, nil
+	}
+
+	// create recovery hook and call it
+	recoveryHook := chunk.NewRecoveryHook(testSender)
+	if err := recoveryHook(ctx, chunkAddr); err != nil {
+		t.Fatal(err)
+	}
+
+	if hookWasCalled != true {
+		t.Fatalf("recovery hook was not called")
+	}
+}
+
+// RecoveryHookTestCase is a struct used as test cases for the TestRecoveryHookCalls func
+type RecoveryHookTestCase struct {
+	name           string
+	ctx            context.Context
+	expectsFailure bool
+}
+
+// TestRecoveryHookCalls verifies that recovery hooks are being called as expected when net store attempts to get a chunk
+func TestRecoveryHookCalls(t *testing.T) {
+	// generate test chunk, store and publisher
+	c := chunktesting.GenerateTestRandomChunk()
+	ref := c.Address()
+	p := "0xbE165fe06c03e4387F79615b7A0b79d535e8D325"
+
+	// test cases variables
+	dummyContext := context.Background() // has no publisher
+	targetContext := context.WithValue(context.Background(), api.TargetsContextKey{}, p)
+
+	for _, tc := range []RecoveryHookTestCase{
+		{
+			name:           "no targets in context",
+			ctx:            dummyContext,
+			expectsFailure: true,
+		},
+		{
+			name:           "targets set in context",
+			ctx:            targetContext,
+			expectsFailure: false,
+		},
```

Comment:
we could also add (maybe for a next iteration) test cases for targets set/unset, but with chunks that are actually retrievable.

that way we could verify that the recovery hooks are _not_ being called, even though they have enough data for that to be the case.

Reply:
This test is just to check if the hook is called or not.. The sending of data is covered in pss and similarly other packages. What you are saying is a integration test that will not have mock packages and not unit test. We should do that in cluster test.

Reply:
> This test is just to check if the hook is called or not.

right, and the hook being called depends on:
- the chunk being missing
- targets being present
- a recovery hook being defined

no?

Reply:
the targets set testcase is there in netstore.. because it is there where we check the targets.

Reply:
> the targets set testcase is there in netstore.. because it is there where we check the targets.

? this is `repair_test.go`